Source/santa-driver/SantaPrefixTree.cc

```diff
@@ -0,0 +1,238 @@
+
+      lck_rw_lock_exclusive_to_shared(spt_lock_);
+    }
+
+    // Get ready for the next iteration.
+    node = (SantaPrefixNode *)node->children[value];
+  }
+
+  if (node_count) *node_count = node_count_;
+
+  lck_rw_unlock_shared(spt_lock_);
+
+  return kIOReturnSuccess;
+}
+
+bool SantaPrefixTree::HasPrefix(const char *string) {
+  lck_rw_lock_shared(spt_lock_);
+
+  auto found = false;
+
+  SantaPrefixNode *node = root_;
+
+  // TODO(bur): Avoid using strlen().
+  size_t len = strlen(string);
+  for (int i = 0; i < len; ++i) {
+    // Only process a byte at a time.
+    uint8_t value = string[i];
+
+    // Get the next char string.
+    node = (SantaPrefixNode *)node->children[value];
+
+    // If it doesn't exist in the tree, no match.
+    if (!node) break;
+
+    // If it does exist, is it a prefix?
+    if (node->isPrefix) {
+      found = true;
+      break;
+    }
+  }
+
+  lck_rw_unlock_shared(spt_lock_);
+
+  return found;
+}
+
+void SantaPrefixTree::Reset(uint32_t *node_count) {
+  lck_rw_lock_exclusive(spt_lock_);
+
+  PruneNode(root_);
+  root_ = new SantaPrefixNode();
+  ++node_count_;
+
+  if (node_count) *node_count = node_count_;
+  LOGD("Prefix tree reset");
+
+  lck_rw_unlock_exclusive(spt_lock_);
+}
+
+void SantaPrefixTree::PruneNode(SantaPrefixNode *target) {
+  if (!target) return;
+
+  // For deep trees a recursive approach will generate too many stack frames.
+  auto stack = OSArray::withCapacity(1);
+  auto seen = OSSet::withCapacity(1);
+
+  // Seed the "stack" with a starting node.
+  auto target_pointer = OSNumber::withNumber((uint64_t)(void *)target, 64);
+  stack->setObject(target_pointer);
+  target_pointer->release();
+
+  // Start at the target node and walk the tree to find all the sub-nodes.
+  while (stack->getCount()) {
+    auto pointer = OSDynamicCast(OSNumber, stack->getLastObject());
+    if (!pointer) {
+      LOGE("Unable complete prune!");
+      break;  // Bail walking the tree, but still delete any seen nodes.
+    }
+
+    seen->setObject(pointer);
+    stack->removeObject(stack->getCount() - 1);
+
+    auto node = (SantaPrefixNode *)pointer->unsigned64BitValue();
+    for (int i = 0; i < 256; ++i) {
+      if (!node->children[i]) continue;
+      auto child_pointer = OSNumber::withNumber((uint64_t)node->children[i], 64);
+      stack->setObject(child_pointer);
+      child_pointer->release();
+    }
+  }
+
+  // Delete all the seen nodes.
+  seen->iterateObjects(^bool(OSObject *object) {
+    auto pointer = OSDynamicCast(OSNumber, object);
+    if (!pointer) {
+      LOGE("Unable delete node!");
+      return false;  // Continue trying to delete the rest of the nodes.
+    }
+    auto node = (SantaPrefixNode *)pointer->unsigned64BitValue();
+    delete node;
+    --node_count_;
+    return false;
+  });
+
+  OSSafeReleaseNULL(stack);
+  OSSafeReleaseNULL(seen);
+}
+
+SantaPrefixTree::~SantaPrefixTree() {
+  // The locking is probably not necessary here, but I am scared to remove it.
```

Comment:
The locking is definitely necessary - while it might not currently be possible for AddPrefix or HasPrefix to be called while this object is being destroyed it's very easy for that to be changed without realizing it could lead to crashes and they may not appear immediately.